File: stacks-queues/stack-ll.c
#include <stdio.h>
#include <stdlib.h>

typedef struct s_node {
	int data;
	struct s_node *next;
} t_node;

typedef struct s_stack {
	t_node *head;
} t_stack;

t_stack *create_stack() {
	t_stack *s = malloc(sizeof(t_stack));
	s->head = NULL;
	return (s);
}

void push_stack(t_stack *s, int data) {
	t_node *node;

	node = malloc(sizeof(t_node));
	node->data = data;
	node->next = s->head;
	s->head = node;
}

int pop_stack(t_stack *s) {
	int value;
	t_node *tmp;

	if (s->head == NULL)
		return -1;
	value = s->head->data;
	tmp = s->head;
	s->head = s->head->next;
	free(tmp);
	return (value);
}

void free_stack(t_stack *s) {
	t_node *tmp;
	t_node *curr;

	curr = s->head;
	while (curr != NULL) {
		tmp = curr;
		curr = curr->next;
		free(tmp);
	}
	free(s);
}

void print_stack(t_stack *s) {
	for (t_node *curr = s->head; curr != NULL; curr = curr->next)
		printf("%d -> ", curr->data);
	printf("NULL\n");
}

int main() {
	t_stack *s;

	s = create_stack();
	print_stack(s);
	push_stack(s, 1);
	push_stack(s, 2);
	push_stack(s, 3);
	print_stack(s);
	pop_stack(s);
	print_stack(s);
	pop_stack(s);
	print_stack(s);
	pop_stack(s);
	print_stack(s);
	pop_stack(s);
	print_stack(s);
	pop_stack(s);
	print_stack(s);
	free_stack(s);
	return (0);
}

File: stacks-queues/stack-ll.js
class Node {
	/**
	 * @param {number} data
	 */
	constructor(data) {
		this.data = data;
		/** @type {Node | null} */
		this.next = null;
	}
}

class Stack {
	constructor() {
		/** @type {Node | null} */
		this.head = null;
	}

	/**
	 * @param {number} data
	 */
	push(data) {
		const node = new Node(data);
		node.next = this.head;
		this.head = node;
	}

	/**
	 * @returns {number}
	 */
	pop() {
		if (this.head == null)
			return ;
		const value = this.head.data;
		this.head = this.head.next;
		return value
	}

	print() {
		for (let curr = this.head; curr != null; curr = curr.next)
			process.stdout.write(`${curr.data} -> `);
		console.log("NULL");
	}
}

const s = new Stack();
s.print();
s.push(1);
s.push(2);
s.push(3);
s.print();
s.pop();
s.print();
s.pop();
s.print();
s.pop();
s.print();
s.pop();
s.print();
s.pop();
s.print();
